lib/ber/reader.js

```diff
@@ -225,25 +226,47 @@ Reader.prototype._readTag = function(tag) {
   if (o === null)
     return null;
 
-  if (this.length > 4)
+  if (this.length > 8)
     throw newInvalidAsn1Error('Integer too long: ' + this.length);
 
   if (this.length > this._size - o)
     return null;
   this._offset = o;
 
-  var fb = this._buf[this._offset];
-  var value = 0;
-
-  for (var i = 0; i < this.length; i++) {
-    value <<= 8;
-    value |= (this._buf[this._offset++] & 0xff);
-  }
+  if (!this.length)
+    return -1;
 
-  if ((fb & 0x80) == 0x80 && i !== 4)
-    value -= (1 << (i * 8));
+  var buffer = new Buffer(this._buf.slice(this._offset, this._offset + this.length));
+  this._offset += this.length;
 
-  return value >> 0;
+  var value;
+  if (this.length > 4) {
+    var highBits = 0;
+    var lowBits = 0;
+    if (this.length == 8) {
+      highBits = buffer.readInt32BE(0);
+      lowBits = buffer.readUInt32BE(4);
+    } else if (this.length == 7) {
+      highBits = buffer.readInt8(0) * Math.pow(2, 16) + buffer.readUInt16BE(1);
+      lowBits = buffer.readUInt32BE(3);
+    } else if (this.length == 6) {
+      highBits = buffer.readInt16BE(0);
+      lowBits = buffer.readUInt32BE(2);
+    } else {
+      highBits = buffer.readInt8(0);
+      lowBits = buffer.readUInt32BE(1);
+    }
+    value = Long.fromBits(lowBits, highBits).toNumber();
+  } else if (this.length == 4) {
+    value = buffer.readInt32BE(0);
+  } else if (this.length == 3) {
+    value = buffer.readInt8(0) * Math.pow(2, 16) + buffer.readUInt16BE(1);
+  } else if (this.length == 2) {
+    value = buffer.readInt16BE(0);
+  } else {
+    value = buffer.readInt8(0);
```

Comment:
I like this a lot. I wonder if it'd be better to be even _more_ explicit (e.g. your `} else {` blocks are implicitly capturing 5 bytes and 1 byte; let's make that explicit)

A switch statement with early return would give you:

```javascript
var highBits;
var lowBits;

switch (this.length) {
  case 0:
    return -1;
  case 1:
    return buffer.readInt8(0);
  case 2:
    return buffer.readInt16BE(0);
  case 3:
    return buffer.readInt8(0) * Math.pow(2, 16) + buffer.readUInt16BE(1);
  case 4:
    return buffer.readInt32BE(0);
  case 5;
    highBits = buffer.readInt8(0);
    lowBits = buffer.readUInt32BE(1);
    return Long.fromBits(lowBits, highBits).toNumber();
  case 6:
    highBits = buffer.readInt16BE(0);
    lowBits = buffer.readUInt32BE(2);
    return Long.fromBits(lowBits, highBits).toNumber();
  case 7:
    highBits = buffer.readInt8(0) * Math.pow(2, 16) + buffer.readUInt16BE(1);
    lowBits = buffer.readUInt32BE(3);
    return Long.fromBits(lowBits, highBits).toNumber();
  case 8:
    highBits = buffer.readInt32BE(0);
    lowBits = buffer.readUInt32BE(4);
    return Long.fromBits(lowBits, highBits).toNumber();
}
```